File: python/tests/test_memory_encoder.py
"""test_memory_encoder.py

Verifies CompanionMemoryEncoder end-to-end: a turn handed to the background
encoder fills the knowledge graph so associative recall can later reach the
episode; attributed beliefs are formed off the hot path (a third party's fact
never becomes the user's); the queue drops rather than blocks when full;
close_async drains remaining work; and an extractor failure is captured, not
fatal. Mirrors the TypeScript pilot (memory_encoder.test.ts) and Go port.
"""
from __future__ import annotations

from typing import Optional

from circle_ai.companion.belief import HeuristicBeliefExtractor, SelfBeliefStore
from circle_ai.companion.memory_encoder import CompanionMemoryEncoder
from circle_ai.memory.extractor import HeuristicKnowledgeGraphExtractor
from circle_ai.memory.graph import (
    InMemoryHippoRagStore,
    InMemoryKnowledgeGraph,
    KnowledgeTriple,
)


class ThrowingExtractor:
    async def extract_from_turn_async(
        self, user_text, assistant_text, source_episode_id, *, ct=None
    ) -> list[KnowledgeTriple]:
        raise RuntimeError("boom")


# ── end-to-end ────────────────────────────────────────────────────────────────


async def test_encodes_a_turn_so_recall_reaches_the_episode_by_a_content_word() -> None:
    graph = InMemoryKnowledgeGraph()
    enc = CompanionMemoryEncoder(HeuristicKnowledgeGraphExtractor(), graph)

    enc.enqueue("I love hiking in Drakensberg", "Sounds wonderful", "ep-hike")
    await enc.close_async()

    assert len(graph.all_triples()) > 0, "graph should have filled from the turn"

    hippo = InMemoryHippoRagStore(graph)
    hits = await hippo.multi_hop_recall_async("drakensberg", 5)
    episode = next((h for h in hits if h.item.id == "ep-hike"), None)
    assert episode is not None, "recall should reach the episode via the extracted edges"
    assert episode.item.text == "I love hiking in Drakensberg"


async def test_forms_attributed_beliefs_off_hot_path_mother_never_user_fact() -> None:
    graph = InMemoryKnowledgeGraph()
    beliefs = SelfBeliefStore()
    enc = CompanionMemoryEncoder(
        HeuristicKnowledgeGraphExtractor(),
        graph,
        HeuristicBeliefExtractor(),
        beliefs,
    )

    enc.enqueue("my mother is diabetic", "Noted", "ep1")
    enc.enqueue("i am vegetarian", "Got it", "ep2")
    await enc.close_async()

    facts = beliefs.self_facts()
    assert not any("diabetic" in f.object for f in facts), (
        "mother's condition must never be a user fact"
    )
    assert any(f.object == "vegetarian" for f in facts)
    assert any(b.object == "diabetic" for b in beliefs.non_self()), (
        "it is still remembered as an audit fact"
    )


# ── queue behaviour ───────────────────────────────────────────────────────────


async def test_drops_writes_beyond_capacity_rather_than_blocking() -> None:
    graph = InMemoryKnowledgeGraph()
    enc = CompanionMemoryEncoder(
        HeuristicKnowledgeGraphExtractor(), graph, None, None, 2
    )

    # Enqueued synchronously before the drain resumes: the 3rd overflows a
    # capacity-2 queue and is dropped.
    enc.enqueue("alpha", "", "e1")
    enc.enqueue("bravo", "", "e2")
    enc.enqueue("charlie", "", "e3")
    await enc.close_async()

    assert graph.get_node("e1") is not None
    assert graph.get_node("e2") is not None
    assert graph.get_node("e3") is None, "the overflow write should have been dropped"


async def test_ignores_an_enqueue_with_a_blank_episode_id() -> None:
    graph = InMemoryKnowledgeGraph()
    enc = CompanionMemoryEncoder(HeuristicKnowledgeGraphExtractor(), graph)
    enc.enqueue("hello", "", "")
    enc.enqueue("hello", "", "   ")
    await enc.close_async()
    assert len(graph.all_triples()) == 0


async def test_captures_an_extractor_failure_without_crashing_the_drain() -> None:
    graph = InMemoryKnowledgeGraph()
    enc = CompanionMemoryEncoder(ThrowingExtractor(), graph)
    enc.enqueue("x", "", "e1")
    await enc.close_async()

    assert isinstance(enc.last_error, Exception)
    assert str(enc.last_error) == "boom"
    # The node was upserted before the extractor ran, so it survives.
    assert graph.get_node("e1") is not None

File: c/tests/test_memory_encoder.c
/*
 * test_memory_encoder.c — CompanionMemoryEncoder end-to-end: a turn fills the
 * graph so associative recall can reach the episode; attributed beliefs formed
 * off the hot path (mother's fact never becomes the user's); overflow dropped;
 * close drains; extractor failure captured, not fatal. Mirrors the Rust suite
 * memory_encoder_test.rs (and TS/Go) 1:1.
 */

#include <stdio.h>
#include <string.h>
#include <assert.h>
#include "circle_ai/circle_ai.h"

/* Throwing extractor: signals failure via *out_count = SIZE_MAX. */
static ca_knowledge_triple_t *throwing_extractor(void *user, const char *u, const char *a,
                                                 const char *src, size_t *out_count) {
    (void)user; (void)u; (void)a; (void)src;
    *out_count = SIZE_MAX;
    return NULL;
}

static bool triples_have_object(ca_knowledge_graph_t *kg, const char *obj) {
    size_t n = 0;
    ca_knowledge_triple_t *all = ca_knowledge_graph_all_triples(kg, &n);
    bool found = false;
    for (size_t i = 0; i < n; ++i) if (strcmp(all[i].object, obj) == 0) { found = true; break; }
    ca_knowledge_triple_free_array(all, n);
    return found;
}

static bool has_node(ca_knowledge_graph_t *kg, const char *id) {
    ca_knowledge_node_t node;
    if (ca_knowledge_graph_get_node(kg, id, &node)) { ca_knowledge_node_free(&node); return true; }
    return false;
}

int main(void) {
    /* ── encodes a turn so associative recall can reach the episode ── */
    {
        ca_knowledge_graph_t *graph = ca_knowledge_graph_create();
        ca_memory_encoder_t *enc = ca_memory_encoder_create(
            ca_kg_extractor_heuristic_adapter, NULL, graph, NULL, NULL, NULL, 0);
        ca_memory_encoder_enqueue(enc, "I love hiking in Drakensberg", "Sounds wonderful", "ep-hike");
        ca_memory_encoder_close(enc);

        size_t n = 0;
        ca_knowledge_triple_t *all = ca_knowledge_graph_all_triples(graph, &n);
        assert(n > 0);
        ca_knowledge_triple_free_array(all, n);

        ca_hippo_store_t *hippo = ca_hippo_store_create(graph);
        ca_memory_hit_t *hits = ca_hippo_store_multi_hop_recall(hippo, "drakensberg", 5, &n);
        bool found = false;
        for (size_t i = 0; i < n; ++i) {
            if (strcmp(hits[i].item.id, "ep-hike") == 0) {
                found = true;
                assert(strcmp(hits[i].item.text, "I love hiking in Drakensberg") == 0);
            }
        }
        assert(found);
        ca_memory_hit_free_array(hits, n);
        ca_hippo_store_destroy(hippo);
        ca_memory_encoder_destroy(enc);
        ca_knowledge_graph_destroy(graph);
    }

    /* ── forms attributed beliefs off the hot path; mother's fact never the user's ── */
    {
        ca_knowledge_graph_t *graph = ca_knowledge_graph_create();
        ca_self_belief_store_t *beliefs = ca_self_belief_store_create();
        ca_memory_encoder_t *enc = ca_memory_encoder_create(
            ca_kg_extractor_heuristic_adapter, NULL, graph,
            ca_belief_extractor_heuristic_adapter, NULL, beliefs, 0);
        ca_memory_encoder_enqueue(enc, "my mother is diabetic", "Noted", "ep1");
        ca_memory_encoder_enqueue(enc, "i am vegetarian", "Got it", "ep2");
        ca_memory_encoder_close(enc);

        size_t fc = 0;
        ca_personal_belief_t *facts = ca_self_belief_store_self_facts(beliefs, &fc);
        bool has_veg = false;
        for (size_t i = 0; i < fc; ++i) {
            assert(strstr(facts[i].object, "diabetic") == NULL);
            if (strcmp(facts[i].object, "vegetarian") == 0) has_veg = true;
        }
        assert(has_veg);
        ca_personal_belief_free_array(facts, fc);

        size_t ac = 0;
        ca_personal_belief_t *audit = ca_self_belief_store_non_self(beliefs, &ac);
        bool has_diab = false;
        for (size_t i = 0; i < ac; ++i) if (strcmp(audit[i].object, "diabetic") == 0) has_diab = true;
        assert(has_diab);
        ca_personal_belief_free_array(audit, ac);

        ca_memory_encoder_destroy(enc);
        ca_self_belief_store_destroy(beliefs);
        ca_knowledge_graph_destroy(graph);
    }

    /* ── drops writes beyond capacity rather than blocking ── */
    {
        ca_knowledge_graph_t *graph = ca_knowledge_graph_create();
        ca_memory_encoder_t *enc = ca_memory_encoder_create(
            ca_kg_extractor_heuristic_adapter, NULL, graph, NULL, NULL, NULL, 2);
        ca_memory_encoder_enqueue(enc, "alpha", "", "e1");
        ca_memory_encoder_enqueue(enc, "bravo", "", "e2");
        ca_memory_encoder_enqueue(enc, "charlie", "", "e3"); /* overflow of capacity-2 */
        ca_memory_encoder_close(enc);

        assert(has_node(graph, "e1"));
        assert(has_node(graph, "e2"));
        assert(!has_node(graph, "e3"));
        ca_memory_encoder_destroy(enc);
        ca_knowledge_graph_destroy(graph);
    }

    /* ── ignores an enqueue with a blank episode id ── */
    {
        ca_knowledge_graph_t *graph = ca_knowledge_graph_create();
        ca_memory_encoder_t *enc = ca_memory_encoder_create(
            ca_kg_extractor_heuristic_adapter, NULL, graph, NULL, NULL, NULL, 0);
        ca_memory_encoder_enqueue(enc, "hello", "", "");
        ca_memory_encoder_enqueue(enc, "hello", "", "   ");
        ca_memory_encoder_close(enc);
        size_t n = 0;
        ca_knowledge_triple_t *all = ca_knowledge_graph_all_triples(graph, &n);
        assert(n == 0);
        ca_knowledge_triple_free_array(all, n);
        ca_memory_encoder_destroy(enc);
        ca_knowledge_graph_destroy(graph);
    }

    /* ── captures an extractor failure without crashing the drain ── */
    {
        ca_knowledge_graph_t *graph = ca_knowledge_graph_create();
        ca_memory_encoder_t *enc = ca_memory_encoder_create(
            throwing_extractor, NULL, graph, NULL, NULL, NULL, 0);
        ca_memory_encoder_enqueue(enc, "x", "", "e1");
        ca_memory_encoder_close(enc);

        const char *last = ca_memory_encoder_last_error(enc);
        assert(last != NULL);
        assert(strcmp(last, "boom") == 0);
        /* Node upserted before the extractor ran, so it survives. */
        assert(has_node(graph, "e1"));
        (void)triples_have_object; /* silence unused in this branch */
        ca_memory_encoder_destroy(enc);
        ca_knowledge_graph_destroy(graph);
    }

    printf("All memory encoder tests passed.\n");
    return 0;
}
